//
// Bentobox
// Proto
// Data types
//

syntax = "proto3";
package bento;

// Wraps data type infomation for all possible kinds of data types
message Type {
    // Defines the data types of the primative kind
    enum Primative {
        INVALID = 0;
        INT32   = 1;
        INT64   = 2;
        FLOAT32 = 3;
        FLOAT64 = 4;
        BOOL    = 5;
        STRING  = 6;
    }

    // Defines a array kind data type that represents an array of primatives
    // Arrays are iterable.
    message Array {
        // The dimensions of the array
        repeated int64 dimensions = 1;
        // The primitive data type that is used to store each element
        Primative element_type = 2;
    }

    oneof kind {
        Primative primitive = 1;
        Array array = 2;
    }
}
